### ISSUE TITLE
`Self` is applied incorrectly during bound generation

### ISSUE THREAD

**Issue by djkoloski** [rkyv/rkyv#188]
Minimal reproduction:

```rust
#[derive(Archive, Deserialize, Serialize)]
pub enum LinkedList<T> {
    Empty,
    Node {
        val: T,
        #[omit_bounds]
        next: Box<Self>,
    }
}
```

The use of `Self` while declaring `next` is causing the wrong type to get bound when generating impls. It's supposed to bound `Box<LinkedList<T>>: Archive` but instead ends up bounding `Box<ArchivedLinkedList<T>>: Archive`. Serde had a similar issue, they fixed it by replacing `Self` with the explicit type during generation.

**Comment by djkoloski** [comment]
Serde's issue: https://github.com/serde-rs/serde/issues/1565

The approach to fix it for rkyv will be similar.

**Comment by djkoloski** [comment]
Fixed by fa49eb1cbb3a93c433935802c68ad72198c9826e
